modules/upgrade/upgrade.go

```diff
@@ -0,0 +1,384 @@
+package upgrade
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/blang/semver"
+	"github.com/rs/zerolog/log"
+
+	"github.com/threefoldtech/zosv2/modules"
+)
+
+type HookType string
+
+const (
+	hookPreCopy   HookType = "pre-copy"
+	hookPostCopy  HookType = "post-copy"
+	hookMigrate   HookType = "migrate"
+	hookPostStart HookType = "post-start"
+)
+
+type UpgradeModule struct {
+	root    string
+	version semver.Version
+	flister modules.Flister
+}
+
+func New(root string, flister modules.Flister) (*UpgradeModule, error) {
+	if err := os.MkdirAll(root, 0770); err != nil {
+		return nil, err
+	}
+
+	var version semver.Version
+	versionPath := filepath.Join(root, "version")
+	version, err := readVersion(versionPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			log.Error().Err(err).Msg("read version")
+			return nil, err
+		}
+		log.Info().Msg("no version found, assuming fresh install")
+		// the file doesn't exist yet. So we are on a fresh system
+		version = semver.MustParse("0.0.1")
+		if err := writeVersion(versionPath, version); err != nil {
+			log.Error().Err(err).Msg("fail to write version")
+			return nil, err
+		}
+	}
+
+	log.Info().Msgf("current version %s", version.String())
+	return &UpgradeModule{
+		version: version,
+		root:    root,
+		flister: flister,
+	}, nil
+}
+
+// FIXME: not sure about the public interface of this package yet
+// most probably will need to run as a daemon too
+func (u *UpgradeModule) Run(period time.Duration, p Publisher) error {
```

Comment:
It seems this function never returns, so this error return can go

Reply:
I'm still actually a bit unhappy about this function.
Don't you think I should maybe remove the infinite loop from the Run method, and move the recurring logic part of the main file ?
So this package only holds logic about upgrade itself. And it's up to the caller to decide how often to run it ?

Reply:
I'm not sure, the period is already decided by an argument. So the question is more, would someone ever want to not call all these actions together I guess?

Reply:
yeah but if I remove the infinite loop, then I also remove the period argument.
And the Run method is actually the only think to run when you want to check/apply a new upgrade.

Reply:
True, I misunderstood your earlier comment, though you wanted to remove this function and just expose the individual functions that it calls. I think that is indeed acceptable.